code/datums/components/pellet_cloud.dm

```diff
@@ -0,0 +1,249 @@
+
+	var/mob/living/shooter /// for if we're an ammo casing being fired
+
+/datum/component/pellet_cloud/Initialize(projectile_type=/obj/item/shrapnel, magnitude=5)
+	if(!isammocasing(parent) && !isgrenade(parent) && !islandmine(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	if(magnitude < 1)
+		CRASH("Invalid magnitude [magnitude] < 1 on pellet_cloud, parent: [parent]")
+		qdel(src, TRUE)
+		return
+
+	src.projectile_type = projectile_type
+
+	if(isammocasing(parent))
+		num_pellets = magnitude
+		if(istype(parent, /obj/item/ammo_casing/caseless))
+			not_done_yet = TRUE
+	else if(isgrenade(parent) || islandmine(parent))
+		radius = magnitude
+		not_done_yet = TRUE
+
+/datum/component/pellet_cloud/Destroy(force, silent)
+	if(not_done_yet)
+		return QDEL_HINT_LETMELIVE
+
+	QDEL_NULL(purple_hearts)
+	QDEL_NULL(pellets)
+	QDEL_NULL(targets_hit)
+	QDEL_NULL(bodies)
+	return ..()
+
+/datum/component/pellet_cloud/RegisterWithParent()
+	if(isammocasing(parent))
+		RegisterSignal(parent, COMSIG_PELLET_CLOUD_INIT, .proc/create_casing_pellets)
+	else if(isgrenade(parent))
+		RegisterSignal(parent, COMSIG_GRENADE_ARMED, .proc/grenade_armed)
+		RegisterSignal(parent, COMSIG_GRENADE_PRIME, .proc/create_blast_pellets)
+	else if(islandmine(parent))
+		RegisterSignal(parent, COMSIG_MINE_TRIGGERED, .proc/create_blast_pellets)
+
+/datum/component/pellet_cloud/UnregisterFromParent()
+	if(not_done_yet)
+		return QDEL_HINT_LETMELIVE
+	UnregisterSignal(parent, list(COMSIG_PELLET_CLOUD_INIT, COMSIG_GRENADE_PRIME, COMSIG_GRENADE_ARMED, COMSIG_MOVABLE_MOVED, COMSIG_MOVABLE_UNCROSSED, COMSIG_MINE_TRIGGERED, COMSIG_ITEM_DROPPED))
+
+/**
+  * create_casing_pellets() is for directed pellet clouds for ammo casings that have multiple pellets (buckshot and scatter lasers for instance)
+  *
+  * Honestly this is mostly just a rehash of [/obj/item/ammo_casing/proc/fire_casing()] for pellet counts > 1, except this lets us tamper with the pellets and hook onto them for tracking purposes.
+  * The arguments really don't matter, this proc is triggered by COMSIG_PELLET_CLOUD_INIT which is only for this really, it's just a big mess of the state vars we need for doing the stuff over here.
+  */
+/datum/component/pellet_cloud/proc/create_casing_pellets(obj/item/ammo_casing/A, atom/target, mob/living/user, fired_from, randomspread, spread, zone_override, params, distro)
+	shooter = user
+	var/targloc = get_turf(target)
+	if(!zone_override)
+		zone_override = shooter.zone_selected
+
+	for(var/i in 1 to num_pellets)
+		A.ready_proj(target, user, SUPPRESSED_VERY, zone_override, fired_from)
+		if(distro)
+			if(randomspread)
+				spread = round((rand() - 0.5) * distro)
+			else //Smart spread
+				spread = round((i / num_pellets - 0.5) * distro)
+
+		RegisterSignal(A.BB, COMSIG_PROJECTILE_SELF_ON_HIT, .proc/pellet_hit)
```

Comment:
Could you use more clear variables? `A.BB` is not very informative.

Reply:
BB is the name ammo casings use for their projectile (presumably BB like a metal ball bearing in a BB gun), would a comment clarifying this suffice since I'm just using vars that were already there

Reply:
At least fix the `A` then. These bad standards proliferate.
`BB` is not kosher either but it's not your fault.